### DISCUSSION TITLE
Adding extra fields to Account model

### DISCUSSION THREAD

**Discussion by mdodell** [janko/rodauth-omniauth#3] [Q&A]
How would I go about adding extra fields to the Account model? I understand that I can use `omniauth_identity_update_hash` after each login. I want to make sure things like the user's name/profile photo are kept to date.

**Comment by janko** [comment]
If I understand correctly, you want to store additional information from the OmniAuth login, but on the `accounts` table instead of the `account_identities` table?

**Reply by janko** [reply]
And also keep them up-to-date with every login, not just store them once on first login?

**Reply by mdodell** [reply]
Yup! Ideally, I'd like to have all of the information I want (i.e. `first_name`, `last_name`, etc.) returned from `rodauth.rails_account`.  Is it better to store that type of information in `account_identities`?"

I'm also operating under the impression that users could use email/password sign-ups to create an account, but then update their information via OAuth. Additionally, if they sign up with OAuth initially, I'd want them to be able to sign in with email/password.

**Reply by mdodell** [reply]
Ideally - keep them up-to-date on every login, in case they change their OAuth profile photo or their name.

**Reply by janko** [reply]
I would recommend using the hook `before_omniauth_create_account` to set initial profile information when the account is registered via OmniAuth login, and not update that information afterwards anymore. You likely want the user to be able to edit their profile information, and with automatic updates you'd overwrite any custom changes the user has made.

If you want this information to be associated to the LinkedIn account and non-modifiable, then it's better to store it on the identity record itself.

And yes, email/password login will still work normally with OmniAuth login, the user just needs to set a password. They should be able to do this via the `change_password` form (they won't be prompted for the current password if there isn't any), and maybe even via `reset_password` flow, I haven't tried that out.

**Reply by mdodell** [reply]
A user can have multiple identity records though, right? I.e. in the case of multiple providers? 

What is the best way to discern where the single source of truth for user information is/what if a user wants to edit their information specifically on my platform (i.e. they want to update only their last name on my platform)? 

My concern is a person calling the `create-account` endpoint with an email they already have via Omniauth provider - I guess in that case, it would probably be rejected since they already have an account via Omniauth?

**Reply by mdodell** [reply]
And if I want to fetch things like the user's first and last name/picture_url - I cannot fetch those from `rodauth.rails_account`, right? Since those would have to be stored under `account_identities` in new fields?

**Reply by mdodell** [reply]
I think looking at the [rodauth-demo-rails](https://github.com/janko/rodauth-demo-rails), my best solution would be to have a `Profile` model that acts as the source of truth for user information, and then my `Account` model has one `Profile` and many `identities`. 

That way, I could get information simply doing `rodauth.rails_account.profile`, and if a user wants to update their information, they can update that model directly. This would probably keep better separation of concerns.

**Reply by janko** [reply]
> A user can have multiple identity records though, right? I.e. in the case of multiple providers?

Yes, if your application offers multiple providers, then they can have multiple identities.

> What is the best way to discern where the single source of truth for user information is/what if a user wants to edit their information specifically on my platform (i.e. they want to update only their last name on my platform)?

If you want the information editable, then you can store them on the `accounts` table or an associated table such as `profiles`. But then you don't want to update that information on every OmniAuth login, because you'd overwrite any custom edits from the user.

The other option is storing it on the `account_identities` table, and making it non-editable and always reflect the latest information by refreshing it on every OmniAuth login.

The option you were suggesting – storing it on the `accounts` table, and making it editable, but still overwriting it on every OmniAuth login – doesn't make sense from the UX standpoint. If the information is editable by the user, then it should never be automatically overwritten.

> My concern is a person calling the create-account endpoint with an email they already have via Omniauth provider - I guess in that case, it would probably be rejected since they already have an account via Omniauth?

That's correct, since the account was already created for that email address.

**Reply by mdodell** [reply]
That makes sense! I'll give this a try - thank you!

**Reply by janko** [reply]
Yes, I generally prefer keeping information not strictly related to authentication in a separate associated table such as `profiles` 👍🏻
